audioipc/src/rpccore.rs

```diff
@@ -0,0 +1,219 @@
+
+use mio::Token;
+
+use crate::ipccore::EventLoopHandle;
+
+// RPC message handler.  Implemented by ClientHandler (for Client)
+// and ServerHandler (for Server).
+pub trait Handler {
+    type In;
+    type Out;
+
+    // Consume a request
+    fn consume(&mut self, request: Self::In) -> Result<()>;
+
+    // Produce a response
+    fn produce(&mut self) -> Result<Option<Self::Out>>;
+}
+
+// Client RPC definition.  This supplies the expected message
+// request and response types.
+pub trait Client {
+    type Request;
+    type Response;
+}
+
+// Server RPC definition.  This supplies the expected message
+// request and response types.  `process` is passed inbound RPC
+// requests by the ServerHandler to be responded to by the server.
+pub trait Server {
+    type Request;
+    type Response;
+
+    fn process(&mut self, req: Self::Request) -> Self::Response;
+}
+
+// RPC Client Proxy implementation
+type ProxyRequest<R, Q> = (R, mpsc::Sender<Q>);
+type ProxyReceiver<R, Q> = mpsc::Receiver<ProxyRequest<R, Q>>;
+
+// Each RPC Proxy `call` returns a blocking waitable ProxyResponse.
+// `wait` produces the response received over RPC from the associated
+// Proxy `call`.
+pub struct ProxyResponse<Q> {
+    inner: mpsc::Receiver<Q>,
+}
+
+impl<Q> ProxyResponse<Q> {
+    pub fn wait(&self) -> Result<Q> {
+        match self.inner.recv() {
+            Ok(resp) => Ok(resp),
+            Err(_) => Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "proxy recv error",
+            )),
+        }
+    }
+}
+
+// RPC Proxy that may be `clone`d for use by multiple owners/threads.
+// A Proxy `call` arranges for the supplied request to be transmitted
+// to the associated Server via RPC.  The response can be retrieved by
+// `wait`ing on the returned ProxyResponse.
+#[derive(Debug)]
+pub struct Proxy<R, Q> {
+    handle: Option<(EventLoopHandle, Token)>,
+    tx: mpsc::Sender<ProxyRequest<R, Q>>,
+}
+
+// XXX: Required for client setup, verify this is safe.
+unsafe impl<R, Q> Sync for Proxy<R, Q> {}
+
+impl<R, Q> Proxy<R, Q> {
+    pub fn call(&self, request: R) -> ProxyResponse<Q> {
+        let (tx, rx) = mpsc::channel();
+        self.tx.send((request, tx)).expect("proxy send error");
+        let (handle, token) = self
+            .handle
+            .as_ref()
+            .expect("proxy not connected to event loop");
+        handle.wake_connection(*token);
+        ProxyResponse { inner: rx }
+    }
+
+    pub(crate) fn connect_event_loop(&mut self, handle: EventLoopHandle, token: Token) {
+        self.handle = Some((handle, token));
+    }
+}
+
+impl<R, Q> Clone for Proxy<R, Q> {
+    fn clone(&self) -> Self {
+        Proxy {
+            handle: self.handle.clone(),
+            tx: self.tx.clone(),
+        }
+    }
+}
+
+fn make_proxy<R, Q>() -> (Proxy<R, Q>, ProxyReceiver<R, Q>) {
+    let (tx, rx) = mpsc::channel();
+    let proxy = Proxy { handle: None, tx };
+    (proxy, rx)
+}
+
```

Comment:
It would be great if we can have some comments stating the relationship between `ClientHandler` and `ServerHandler`:
```
ClientHandler       ServerHandler
    produce ----------> consume 
    consume <---------- produce
```